openquake/hazardlib/near_fault.py

```diff
@@ -0,0 +1,434 @@
+    segment_s = [node0_xyz, node1_xyz, node2_xyz, node3_xyz]
+    segment_e = [node1_xyz, node2_xyz, node3_xyz, node0_xyz]
+
+    # set buffering bu
+    buf = 0.0001
+    atol = 0.0001
+
+    loop = True
+    exit_flag = False
+    looptime = 0.
+    while loop:
+        x_min = np.min(np.array([node0_xyz[0], node1_xyz[0], node2_xyz[0],
+                                node3_xyz[0]])) - buf
+        x_max = np.max(np.array([node0_xyz[0], node1_xyz[0], node2_xyz[0],
+                                node3_xyz[0]])) + buf
+        y_min = np.min(np.array([node0_xyz[1], node1_xyz[1], node2_xyz[1],
+                                node3_xyz[1]])) - buf
+        y_max = np.max(np.array([node0_xyz[1], node1_xyz[1], node2_xyz[1],
+                                node3_xyz[1]])) + buf
+        n_seg = 0
+        exit_flag = False
+        for (seg_s, seg_e) in zip(segment_s, segment_e):
+
+            pa = np.array([seg_s, pp_xyz])
+            pb = np.array([seg_e, hypocenter_xyz])
+
+            si = pb - pa
+
+            ni = si / np.power(
+                np.dot(np.sum(si ** 2, axis=1).reshape(2, 1),
+                       np.ones((1, 3))), 0.5)
+
+            nx = ni[:, 0].reshape(2, 1)
+            ny = ni[:, 1].reshape(2, 1)
+            nz = ni[:, 2].reshape(2, 1)
+            sxx = np.sum(nx ** 2 - 1)
+            syy = np.sum(ny ** 2 - 1)
+            szz = np.sum(nz ** 2 - 1)
+            sxy = np.sum(nx * ny)
+            sxz = np.sum(nx * nz)
+            syz = np.sum(ny * nz)
+            s = np.array([sxx, sxy, sxz, sxy, syy, syz, sxz, syz,
+                         szz]).reshape(3, 3)
+
+            cx = np.sum(pa[:, 0].reshape(2, 1) * (nx ** 2 - 1) +
+                        pa[:, 1].reshape(2, 1) * [nx * ny] +
+                        pa[:, 2].reshape(2, 1) * (nx * nz))
+            cy = np.sum(pa[:, 0].reshape(2, 1) * [nx * ny] +
+                        pa[:, 1].reshape(2, 1) * [ny ** 2 - 1] +
+                        pa[:, 2].reshape(2, 1) * [ny * nz])
+            cz = np.sum(pa[:, 0].reshape(2, 1) * [nx * nz] +
+                        pa[:, 1].reshape(2, 1) * [ny * nz] +
+                        pa[:, 2].reshape(2, 1) * [nz ** 2 - 1])
+            c = np.array([cx, cy, cz]).reshape(3, 1)
+            p_intersect = np.linalg.solve(s, c)
+
+            vector1 = (p_intersect.flatten() - hypocenter_xyz) / \
+                sum((p_intersect.flatten() - hypocenter_xyz) ** 2) ** 0.5
+            vector2 = (pp_xyz - hypocenter_xyz) / \
+                sum((pp_xyz - hypocenter_xyz) ** 2) ** 0.5
+            vector3 = (np.array(seg_e) - np.array(seg_s)) / \
+                sum((np.array(seg_e) - np.array(seg_s)) ** 2) ** 0.5
+            vector4 = (p_intersect.flatten() - np.array(seg_s)) / \
+                sum((p_intersect.flatten() - np.array(seg_s)) ** 2) ** 0.5
+
+            ppph = dst.pdist([pp, hypocenter_xyz])
+            pdph = dst.pdist([p_intersect.flatten(), hypocenter_xyz])
+            n_seg = n_seg + 1
+
```

Comment:
This function is too long. I would think about breaking it in pieces. Helper functions are welcome.